lib/chat_transcript_service.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+
+##
+# Used to generate BBCode [chat] tags for the message IDs provided.
+#
+# If there is > 1 message then the channel name will be shown at
+# the top of the first message, and subsequent messages will have
+# the chained attribute, which will affect how they are displayed
+# in the UI.
+#
+# Subsequent messages from the same user will be put into the same
+# tag. Each new user in the chain of messages will have a new [chat]
+# tag created.
+#
+# A single message will have the channel name displayed to the right
+# of the username and datetime of the message.
+class ChatTranscriptService
+  CHAINED_ATTR = "chained=\"true\""
+  MULTIQUOTE_ATTR = "multiQuote=\"true\""
+
+  class ChatTranscriptBBCode
+    attr_reader :channel, :multiquote, :chained
+
+    def initialize(
+      channel: nil,
+      multiquote: false,
+      chained: false
+    )
+      @channel = channel
+      @multiquote = multiquote
+      @chained = chained
+      @messages = []
+    end
+
+    def <<(message)
+      @messages << message
+    end
+
+    def render
+      attrs = [quote_attr(@messages.first)]
+      attrs << channel_attr if channel
+      attrs << MULTIQUOTE_ATTR if multiquote
+      attrs << CHAINED_ATTR if chained
+
+      <<~MARKDOWN
+      [chat #{attrs.join(" ")}]
+      #{@messages.map(&:message).join("\n\n")}
+      [/chat]
+      MARKDOWN
+    end
+
+    def quote_attr(message)
+      "quote=\"#{message.user.username};#{message.id};#{message.created_at.iso8601}\""
+    end
+
+    def channel_attr
+      "channel=\"#{channel.name}\""
+    end
+  end
+
+  def initialize(channel, message_ids)
+    @channel = channel
+    @message_ids = message_ids
+  end
+
+  def generate_bbcode
```

Comment:
I wonder if we should name this something more generic. `bbcode` is an internal implementation detail which may or may not change in the future. If it does change, might be annoying to  have to update all the references everywhere. Maybe something like `generate_markdown`?

Reply:
Yeah good point I was wondering about that a bit, will change